Detecting infinite dependency loops attempts

AndreyDobrovolskyOdessa opened the pull request:
Achieved with the help of storing the whole dependency path in REDO_TRACK
environment variable and searching the next target full path inside REDO_TRACK.
If found means an attempt for infinite loop, required branch is ommitted.
If not found, target full path is appended to REDO_TRACK and .do file is
envoked with the updated REDO_TRACK value.

AndreyDobrovolskyOdessa committed:
Detecting infinite dependemcy loops attempts

Achieved with the help of storing the whole dependency path in REDO_TRACK
environment variable and searching the next target full path inside REDO_TRACK.
If found means an attempt for infinite loop, required branch is ommitted.
If not found, target full path is appended to REDO_TRACK and .do file is
envoked with the updated REDO_TRACK value.

AndreyDobrovolskyOdessa commented:
Hi, Leah!
In redo.c infinite dependencies are marked as the bug. The current version is locking on such events. This request is proposal of the simple technique how to deal with them. In my implementation attempts to loop infinitely are detected and ignored - .do file is skipped and appearance of target name in the .dep file is blocked. The whole build process isn't interrupted, warning is thrown into stderr. Maybe exiting with some error code will be more appropriate.
Best regards!

leahneukirchen commented:
Hm, does this work given a parallel build?

AndreyDobrovolskyOdessa commented:
Depends on what You mean as "does work". This tool process local data, without colllecing and processing information about the whole graph. So it can warn You about loops present in the requested graph topology, described by .do files, avoid locking and wipe out some branches from the .dep files to avoid infinite recursion during check_deps(). Please take into consideration, that in case loops present, resulting .dep-described topology will differ from the topology requested by the .do files collection. And if for the single build You got "loop warning" means that something is wrong with Your .do files content, and starting parallel build on such graph could hardly produce something of better quality :)
In general if You will create looped dependency graph of .do files and start parallel builds on it without any tests, You will likely got into the mutual lock, and to resolve such a lock You will need some global superviser, which can collect data from all builds running and running. The complexity of such superviser can easily overcome the current redo.c overall  complexity imho.

The minimal example is:
 
cat > a.do << EOF
redo-ifchange b
cat b > $3
EOF

cat > b.do << EOF
redo-ifchange c
cat c > $3
EOF

cat > c.do << EOF
redo-ifchange a
cat a > $3
EOF

and started simultaneously (let's imagine) redo a, redo b, redo c will lock each other and the hypothetical superviser able to resolve such a lock will need information from and control over all of these builds simultaneously, not even any pair.

Can You please describe what is preferred behaviour of the hypothetical "loop resolver" for the above example?

leahneukirchen commented:
Yes, after some reflection I think the approach is okay. I'll review the code tomorrow.

jgarvin commented:
Is the whole dependency path guaranteed to fit inside an environment variable? On Linux command line arguments and environment variables combined must fit inside a [small limit](https://devblogs.microsoft.com/oldnewthing/20100203-00/?p=15083#:~:text=The%20theoretical%20maximum%20length%20of,a%20limit%20of%2032767%20characters). Also this sounds quadratic -- at each layer doing a longer and longer linear search of an ever growing string.

leahneukirchen commented:
The Linux limit is far higher, I think 25% of the stack or something. So it can store 100 entries for sure. And I doubt reasonable build depths exceed something like 10...

AndreyDobrovolskyOdessa committed:
append_branch(): testing pattern end

Avoid false match triggers on the targets, matching the search
pattern, but having additional symbols tailing.

AndreyDobrovolskyOdessa commented:
Hi Leah!

You wrote:
> Hm, does this work given a parallel build?

It does, but only for lock-free paralellizing.

You wrote:
> The Linux limit is far higher, I think 25% of the stack or something. So it can store 100 entries for sure. And I doubt reasonable build depths exceed something like 10...

Even more. My version of redo-c is limited by the default number of the files files opened by the process (1024) so it refuses to dive deeper than 510 nesting - built without optimizations. If built with optimizations default stack is exhausted approx. at 480 nesting level.
